src/agora/dozf.cc

```diff
@@ -223,34 +242,105 @@ void DoZF::ZfTimeOrthogonal(size_t tag) {
-      arma::cx_fmat calib_ul_mat_prev(
+      arma::cx_fmat pre_calib_ul_msum_mat(
           reinterpret_cast<arma::cx_float*>(
-              calib_ul_buffer_[frame_cal_slot_prev]),
+              calib_ul_msum_buffer_[frame_cal_slot_prev]),
           cfg_->OfdmDataNum(), cfg_->BfAntNum(), false);
-      arma::cx_fvec calib_dl_vec =
-          (calib_dl_mat.row(cur_sc_id) + calib_dl_mat_prev.row(cur_sc_id)).st();
-      arma::cx_fvec calib_ul_vec =
-          (calib_ul_mat.row(cur_sc_id) + calib_ul_mat_prev.row(cur_sc_id)).st();
-      calib_vec = calib_dl_vec / calib_ul_vec;
+
+      // Calculate a moving sum
+      cur_calib_dl_msum_mat.row(cur_sc_id) =
+          cur_calib_dl_mat.row(cur_sc_id) +
+          pre_calib_dl_msum_mat.row(cur_sc_id) -
+          old_calib_dl_mat.row(cur_sc_id);
+      cur_calib_ul_msum_mat.row(cur_sc_id) =
+          cur_calib_ul_mat.row(cur_sc_id) +
+          pre_calib_ul_msum_mat.row(cur_sc_id) -
+          old_calib_ul_mat.row(cur_sc_id);
+
+      if (cfg_->InitCalibRepeat() == 0u && frame_grp_id == 0)
+        // fill with one until one full sweep
+        // of  calibration data is done
+        calib_vec.fill(arma::cx_float(1, 0));
+      else
+        calib_vec = (cur_calib_dl_msum_mat.row(cur_sc_id) /
+                     cur_calib_ul_msum_mat.row(cur_sc_id))
+                        .st();
+
+      if (kRecordCalibrationMats == true) {
+        if (cfg_->Frame().IsRecCalEnabled() && (frame_id >= TX_FRAME_DELTA)) {
+          // Write to file when a full sweep of calibration in done
+          if ((frame_id - TX_FRAME_DELTA) % cfg_->AntGroupNum() ==
+                  cfg_->AntGroupNum() - 1 &&
+              cur_sc_id == 0) {
+            Utils::SaveVec(cur_calib_dl_mat.row(cur_sc_id).st(),
+                           "calib_dl_vec.m",
+                           "online_calib_dl_vec" +
+                               std::to_string(frame_id / cfg_->AntGroupNum()),
+                           true /*append*/);
+            Utils::SaveVec(cur_calib_ul_mat.row(cur_sc_id).st(),
+                           "calib_ul_vec.m",
+                           "online_calib_ul_vec" +
+                               std::to_string(frame_id / cfg_->AntGroupNum()),
+                           true /*append*/);
+
+            arma::cx_fvec calib_vec_inst = (cur_calib_dl_mat.row(cur_sc_id) /
+                                            cur_calib_ul_mat.row(cur_sc_id))
+                                               .st();
+            Utils::SaveVec(calib_vec_inst, "calib_vec.m",
+                           "online_calib_vec_inst" +
+                               std::to_string(frame_id / cfg_->AntGroupNum()),
+                           true /*append*/);
+            Utils::SaveVec(calib_vec, "calib_vec.m",
+                           "online_calib_vec_mean" +
+                               std::to_string(frame_id / cfg_->AntGroupNum()),
+                           true /*append*/);
+          }
+        }
+      }
```

Comment:
Can we modularize this part of code (separating it into smaller functions)? This function too long now and this part of code might need to be applied to frequency orthogonal case in the future.